[SYCL][BLAS] Move portBLAS to UXL Foundation as generic SYCL BLAS

s-Nick opened the pull request:
# Description

This PR aims to keep oneMath portBLAS backend compatible and usable after the upcoming changes to it. Everything will be moved to UXL Foundation as a generic SYCL BLAS implementation. 
Here all occurences to portBLAS are removed and renamed to a more appropriate name.

# Checklist

## All Submissions

- [x] Do all unit tests pass locally? Attach a log.
[intel_igpu_log.txt](https://github.com/user-attachments/files/18092542/intel_igpu_log.txt)

s-Nick committed:
Update CMake configuration to new name

First changes to CMake config to move from portBLAS to oneMath generic
blas kernels.

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

s-Nick committed:
Change namespace

Update namespace from portBLAS to generic.

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

s-Nick committed:
Rename and move files

Moved all files portblas related to new `generic` name. Moved directory
path to `generic`.

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

s-Nick committed:
Update include files, namespace and function names

Update included headers to new name after removing portblas.
Update namespace from portblas to generic.
Functions that had portblas in their name have been renamed to use
`generic` instead.

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

s-Nick committed:
Update generic blas backend CMakeLists

s-Nick committed:
Update macros

Remove PORTBLAS from macros in favor of GENERIC_BLAS

s-Nick committed:
Update documentations and comments

Remove portBLAS from documentations and comments

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

s-Nick committed:
Fix after rebasing on oneMath renaming branch

Fix renaming part from portBLAS to generic after rebased.

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

s-Nick committed:
Update README to reference new project name

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

s-Nick commented on README.md:
```diff
@@ -65,7 +65,7 @@ oneMath is part of the [UXL Foundation](http://www.uxlfoundation.org).
         </tr>
         <tr>
-            <td align="center"><a href="https://github.com/codeplaysoftware/portBLAS"> portBLAS </a></td>
+            <td align="center"><a href="https://github.com/codeplaysoftware/portBLAS"> generic SYCL BLAS </a></td>
```
There many links that still mention and bring to portBLAS codebase. I will update them as soon as the migration is completed.

s-Nick replied:
Updated in [2abd2c7](https://github.com/uxlfoundation/oneMath/pull/618/commits/2abd2c7eb9ba7a90e0f93c8166c3df8fe5652ec4)

s-Nick committed:
Fix formatting

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

Rbiessy commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -236,18 +236,18 @@ SYCL enables portable heterogeneous computing on a wide range of accelerators.
-Pure SYCL backends: portBLAS and portFFT
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Pure SYCL backends: generic BLAS and portFFT
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
```suggestion
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It's missing 2 to underline the whole title

Rbiessy commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
 
-Building for portBLAS
+Building for oneMath generic SYCL BLAS
 ---------------------
```
```suggestion
Building for oneMath generic SYCL BLAS
--------------------------------------
```

Rbiessy commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
+By default, the generic BLAS backend is not tuned for any specific device.
 This tuning is required to achieve best performance.
-portBLAS can be tuned for a specific hardware target by adding compiler
+OneMath_sycl_blas can be tuned for a specific hardware target by adding compiler
```
```suggestion
OneMath SYCL BLAS can be tuned for a specific hardware target by adding compiler
```
I think this is more consistent with the rest of the document?

Rbiessy commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
 for more details.
 
-Both portBLAS and portFFT are used as header-only libraries, and will be
+Both onemath_sycl_blas and portFFT are used as header-only libraries, and will be
```
```suggestion
Both oneMath SYCL BLAS and portFFT are used as header-only libraries, and will be
```

Rbiessy commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -59,7 +59,7 @@ or ``clang++`` and ``clang`` respectively when using the Open DPC++ Compiler.
 for each desired backend. By default, only the ``MKLGPU`` and ``MKLCPU``
 backends are enabled. Multiple backends for multiple device vendors can be
-enabled at once (albeit with limitations when using portBLAS and portFFT). The
+enabled at once (albeit with limitations when using onemath_sycl_blas and portFFT). The
```
```suggestion
enabled at once (albeit with limitations when using oneMath SYCL BLAS and portFFT). The
```

s-Nick committed:
Apply suggestions from code review

Co-authored-by: Romain Biessy <romain.biessy@codeplay.com>

andrewtbarker commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -59,7 +59,7 @@ or ``clang++`` and ``clang`` respectively when using the Open DPC++ Compiler.
 for each desired backend. By default, only the ``MKLGPU`` and ``MKLCPU``
 backends are enabled. Multiple backends for multiple device vendors can be
-enabled at once (albeit with limitations when using portBLAS and portFFT). The
+enabled at once (albeit with limitations when using oneMath SYCL BLAS and portFFT). The
```
Should this be "generic SYCL BLAS" or "oneMath generic SYCL BLAS"?

andrewtbarker commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -236,18 +236,18 @@ SYCL enables portable heterogeneous computing on a wide range of accelerators.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-`portBLAS <https://github.com/codeplaysoftware/portBLAS>`_ and `portFFT
+`Generic BLAS <https://github.com/codeplaysoftware/portBLAS>`_ and `portFFT
```
```suggestion
`Generic SYCL BLAS <https://github.com/codeplaysoftware/portBLAS>`_ and `portFFT
```

andrewtbarker commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -435,11 +435,11 @@ Build oneMath for the BLAS domain on a generic SYCL device:
-portBLAS backend only, for a generic SYCL device supported by the Open DPC++
-project.
+generic BLAS backend only, for a generic SYCL device supported by the
+Open DPC++ project.
```
What is the "Open DPC++ project"? I recommend ending the sentence after "for a generic SYCL device."

andrewtbarker commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
+By default, the generic BLAS backend is not tuned for any specific device.
 This tuning is required to achieve best performance.
-portBLAS can be tuned for a specific hardware target by adding compiler
+OneMath SYCL BLAS can be tuned for a specific hardware target by adding compiler
```
```suggestion
The generic BLAS backend can be tuned for a specific hardware target by adding compiler
```

andrewtbarker commented on docs/building_the_project_with_dpcpp.rst:
```diff
@@ -257,35 +257,35 @@ experimental and currently only support a subset of operations and features.
 for more details.
 
-Both portBLAS and portFFT are used as header-only libraries, and will be
+Both oneMath SYCL BLAS and portFFT are used as header-only libraries, and will be
```
```suggestion
Both generic SYCL BLAS and portFFT are used as header-only libraries, and will be
```

andrewtbarker reviewed (commented):
Generally looks good, I think the use of names in the documentation could be a bit more consistent.

One question (which probably should have been brought up earlier in this process, sorry): "generic BLAS" to me suggests any library that implements the BLAS standard (eg ESSL, AOCL, NVPL, BLIS, ...) but that is not what is meant here. So I guess the suggestion is to include "SYCL" a lot for clarity.

s-Nick committed:
Address PR comments

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

s-Nick commented:
Thank you for your review @andrewtbarker 
I understand what you mean, I'll give another look to documentation for name consistency. In the mean time, I addressed all your suggestions in [b88b452](https://github.com/uxlfoundation/oneMath/pull/618/commits/b88b4527da5a32fe7a6529ebbb94704a12b1a7a1)

s-Nick committed:
Update CMakeLists to point to the new UXL Foundation repository

s-Nick committed:
Update documentation links to point to the new ones

Rbiessy reviewed (approved):
Thanks LGTM, if there are no more concerns I will merge this tomorrow morning!

Rbiessy commented on CMakeLists.txt:
```diff
@@ -45,7 +45,7 @@ endif()
 option(ENABLE_ROCBLAS_BACKEND "Enable the rocBLAS backend for the BLAS interface" OFF)
 option(ENABLE_NETLIB_BACKEND "Enable the Netlib backend for the BLAS interface" OFF)
-option(ENABLE_PORTBLAS_BACKEND "Enable the portBLAS backend for the BLAS interface. Cannot be used with other BLAS backends." OFF)
+option(ENABLE_GENERIC_BLAS_BACKEND "Enable the generic BLAS backend for the BLAS interface. Cannot be used with other BLAS backends." OFF)
```
Can you add some logic below to check if the user defines `ENABLE_PORTBLAS_BACKEND`? I suggest that we add a warning message explaining is `ENABLE_PORTBLAS_BACKEND` deprecated. We can automatically set `ENABLE_GENERIC_BLAS_BACKEND=ON` in this case to keep the internal CI working.

Rbiessy commented on src/blas/backends/generic/CMakeLists.txt:
```diff
@@ -17,11 +17,11 @@
+set(LIB_NAME onemath_blas_generic)
 set(LIB_OBJ ${LIB_NAME}_obj)
 
-if(NOT DEFINED PORTBLAS_TUNING_TARGET)
```
Similarly to my comment, can we warn the user but keep things working if `PORTBLAS_TUNING_TARGET` is set?

s-Nick replied:
I added a warning message for both in  [e8d22aa](https://github.com/uxlfoundation/oneMath/pull/618/commits/e8d22aac5b8e6cd0d2d7e69aa0e1412eb193b79c)

Rbiessy reviewed (changes requested):
I noticed the GitHub CI is not running the portBLAS tests anymore as it still uses the portBLAS options. Can you rename the occurrences in https://github.com/uxlfoundation/oneMath/blob/develop/.github/workflows/pr.yml#L37?

s-Nick committed:
Add CMake WARNING message in case old definitions are used

In case definition related to portBLAS are passed during configuration
emits a warning message, convert the flag to new one and continue with
proper configuration.

Signed-off-by: nscipione <nicolo.scipione@codeplay.com>

s-Nick committed:
Merge branch 'develop' of github.com:uxlfoundation/oneMath into pb_to_blas_kernel

s-Nick committed:
Update GitHub workflows

s-Nick commented:
@Rbiessy I updated the workflows in [a97bcb2](https://github.com/uxlfoundation/oneMath/pull/618/commits/a97bcb2d686391bb14c58f0b05b5412591af2ece) and it is running the CI now

zettai-reido commented:
Looks approachable.
Do we really need those pre-C++17 namespaces in cpp files instead of
`onemath::blas::generic` style?

s-Nick commented:
Thank you for your feedback @zettai-reido.
There is not particular reason for that namespace a part from following the contributing guideline (NS4) https://github.com/uxlfoundation/oneMath/blob/develop/CONTRIBUTING.md#ns-namespaces
I think this change is outside the scope of this PR. If we  want to update the style we would need to update the documentation and other backend. Could you open an issue for it?

andreyfe1 reviewed (approved):
Looks good to me overall. Thanks!
